cachi2/core/package_managers/metayarn.py

```diff
@@ -0,0 +1,122 @@
+import logging
+import re
+from collections import defaultdict
+from functools import partial
+from pathlib import Path
+from typing import Iterable, Optional, Union
+
+from cachi2.core.errors import PackageRejected
+from cachi2.core.models.input import PackageInput, Request
+from cachi2.core.models.output import RequestOutput
+from cachi2.core.package_managers import yarn, yarn_classic
+from cachi2.core.resolver import _merge_outputs
+from cachi2.core.rooted_path import RootedPath
+
+log = logging.getLogger(__name__)
+
+
+def _get_path_to_yarn_lock(
+    package: PackageInput,
+    source_dir: RootedPath,
+) -> Path:
+    """Construct a path to package's lockfile.
+
+    Raise an exception when there is no lockfile.
+    """
+    yarnlock_path = source_dir.join_within_root("yarn.lock")
+    if yarnlock_path.path.exists():
+        return yarnlock_path.path
+
+    raise PackageRejected(f"Yarn lockfile is missing in {package.path}", solution=None)
```

Comment:
I understand and you're right. I'd just like to avoid having the same check at multiple places, it just makes debugging a bit harder (definitely not impossible). Like I mentioned in another comment I raise, this meta module should be as simple as possible, and so if "yarn lockfile v1" isn't matched, I'd default to processing everything with yarn v2+ since that has this check and is raising this exception already.

Reply:
> if we could be sure that no one would ever mix different Yarn versions. Since mixing them is possible now I opted to preserving the behavior. This resulted in the need to separate potentially mixed dependencies. If this is not what we want then yes, most of this code could be scraped and replaced with the one method above.

True. I still think though that some of the proposed logic would be better hosted in the respective yarn backend modules we have, maybe even paired with more thourough checks we already have in place in those modules when it comes to project files and just let the given parser handle the lockfile - this would actually be a good field test of our yarn backends.

Reply:
Yes, concrete Yarn backends are definitely a better place for that. I have just realized that I can achieve the same results by lifting the check to corresponding PM and splitting packages list into individual packages here before feeding them to PMs instead of separating them by Yarn version and then sending to handlers. I'll push the next change with it (as a separate commit while this is a draft, will squash later).